server/src/main/java/org/candlepin/controller/SuspendModeTransitioner.java

```diff
@@ -0,0 +1,232 @@
+
+        if (maxDelay != -1 && delay > maxDelay) {
+            log.debug("Maximum delay {} reached", maxDelay);
+            delay = maxDelay;
+        }
+
+        log.debug("Next Transitioner check will run after {} seconds", delay);
+        if (failedAttempts > 3) {
+            log.info("SuspendModeTransitioner failed to reconnect to the Qpid Broker " +
+                "{} times, backing off with next reconnect {} seconds", failedAttempts,
+                delay);
+        }
+        execService.schedule(this, delay, TimeUnit.SECONDS);
+    }
+
+    @Override
+    public void run() {
+        log.debug("Executing periodic transition attempt");
+        try {
+            transitionAppropriately();
+        }
+        finally {
+            /**
+             * The Transitioner is periodic task. But we want to have flexibility of setting
+             * different delays between executions. As the amount of failed atttempts to connect
+             * rise, we want to also increase the delay. Scheduling in finally seems to be the
+             * easiest way to achieve that.
+             */
+            schedule();
+        }
+    }
+
+    /**
+     * Attempts to transition Candlepin according to current Mode and current status of
+     * the Qpid Broker. Logs and swallows possible exceptions - theoretically
+     * there should be none.
+     *
+     * Most of the time the transition won't be required and this method will be no-op.
+     * There is an edge-case when transitioning from SUSPEND to NORMAL mode.
+     * During that transition, there is a small time window between checking the
+     * Qpid status and attempt to reconnect. If the Qpid status is reported as
+     * Qpid up, the transitioner will try to reconnect to the broker. This reconnect
+     * may fail. In that case the transition to NORMAL mode shouldn't go through
+     */
+    public synchronized void transitionAppropriately() {
+        log.debug("Attempting to transition to appropriate Mode");
+        try {
+            QpidStatus status = qmf.getStatus();
+            CandlepinModeChange modeChange = modeManager.getLastCandlepinModeChange();
+
+            log.debug("Qpid status is {}, the current mode is {}", status, modeChange);
+
+            if (status != QpidStatus.CONNECTED) {
+                qpidConnection.setConnectionStatus(STATUS.JMS_OBJECTS_STALE);
+            }
+
+            if (modeChange.getMode() == Mode.SUSPEND) {
+                switch (status) {
+                    case CONNECTED:
+                        log.info("Connection to qpid is restored! Reconnecting Qpid and" +
+                            " Entering NORMAL mode");
+                        failedAttempts = 0;
+                        modeManager.enterMode(Mode.NORMAL, Reason.QPID_UP);
+                        break;
+                    case FLOW_STOPPED:
+                    case DOWN:
+                        failedAttempts++;
```

Comment:
There is no upper bound on failedAttempts, with poor configuration and a very negligent admin this will overflow and start counting down negative numbers.